### PR TITLE
ENG-89314: dynamic SchemaNamePrefix support in clio MCP and tools

### PR DESCRIPTION
**Pull request by m-dymytrova**
## Overview

Creatio environments can be configured with any `SchemaNamePrefix` — `"Usr"`, `"My"`, an empty string, or anything else. Before this PR, clio had two problems:

1. **`create-app` silently fell back to `"Usr"`** when `SchemaNamePrefix` could not be read from the environment, causing schemas to be named with the wrong prefix on non-default environments without any indication that something went wrong.
2. **`get-app-info` never surfaced the prefix at all**, leaving agents that work with existing apps with no reliable way to discover which prefix to use before naming entities, columns, lookups, and pages.

This PR introduces a `get-schema-name-prefix` MCP tool as an explicit, dedicated prefix discovery mechanism, wires `SchemaNamePrefix` into `create-app`'s response so agents can read it immediately after app creation, then extends the same to `get-app-info` so existing-app workflows are covered without a separate tool call. The silent `"Usr"` fallback is removed — prefix read failures now propagate as explicit errors. MCP tool descriptions across entity, page, and sync tools are updated to guide agents to use the active prefix consistently.

**Design principle:** the agent owns schema naming; clio provides the prefix as discoverable context; failures reading the prefix must be explicit.

---

### New MCP tool: `get-schema-name-prefix`
- Reads `SchemaNamePrefix` from the target environment
- Returns `success: false` with an error message on failure — no silent fallback
- Intended for flows where the prefix is needed before `create-app` or `get-app-info`

### `create-app`: dynamic prefix applied automatically
- `ApplicationCreateService` injects `ISysSettingsManager` factory and reads `SchemaNamePrefix` before code generation
- `SanitizeCode`, `GenerateCodeFromName`, `DeriveNameFromCode` accept and apply the active prefix
- Already-prefixed codes are accepted without duplication
- `schema-name-prefix` is returned in the `create-app` response so agents can use it for all subsequent schema naming in the session
- Silent fallback to `"Usr"` removed — read failures propagate as explicit errors

### `get-app-info`: prefix now included in response
- `ApplicationInfoService` injects `ISysSettingsManager` factory and reads `SchemaNamePrefix`
- `schema-name-prefix` is included in the `get-app-info` response, covering existing-app workflows without a separate `get-schema-name-prefix` call

### MCP tool descriptions updated
- `EntitySchemaTool`: `schema-name`, `columns`, and column `name` descriptions include prefix guidance
- `PageCreateTool`: `schema-name` description includes prefix guidance
- `SchemaSyncTool`: `schema-name` and `columns` descriptions include prefix guidance
- `SchemaNamePrefixTool`: description updated to note that both `create-app` and `get-app-info` already return the prefix

### Guidance resources
- `AppModelingGuidanceResource`: updated with dynamic prefix discovery flow
- `ExistingAppMaintenanceGuidanceResource`: directs agents to capture `schema-name-prefix` from `get-app-info` before naming any schema artifact in the session

### CLI documentation
- `create-app.md`, `create-app.txt`: document auto-applied prefix behaviour and no-double-prefix guarantee

---

## Tests

**Unit:**
- `ApplicationCreateServiceTests`: 7 new tests for non-default prefix, no-double-prefix, empty prefix, digit-start rejection, and exception propagation (replaces fallback-to-Usr test)
- `ApplicationInfoServiceTests`: 3 new tests for prefix passthrough, empty prefix, and exception propagation; constructor updated for new factory parameter
- `ApplicationToolTests`, `ToolContractGetToolTests`: assertions updated for `schema-name-prefix` field

**E2E:**
- `SchemaNamePrefixToolE2ETests`: validates `get-schema-name-prefix` against a live environment
- `ApplicationToolE2ETests`: asserts `schema-name-prefix` is non-null in `get-app-info` response
- `ToolContractGetToolE2ETests`: asserts `schema-name-prefix` present in `get-app-info` and `create-app` output contracts

### CODE REVIEW

File: clio/Command/McpServer/Tools/ToolContractGetTool.cs [hunk 1]
```diff
@@ -594,6 +594,7 @@ private static ToolContractDefinition BuildApplicationCreate() {
 				Field(ApplicationVersionFieldName, StringType, InstalledApplicationVersionDescription),
 				Field("entities", ArrayType, "Application entities."),
 				Field(PagesFieldName, ArrayType, "Primary-package Freedom UI pages using list-pages item shape (`schema-name`, `uId`, `packageName`, `parentSchemaName`)."),
+				Field("schema-name-prefix", StringType, "Active SchemaNamePrefix resolved from the environment. Use as the prefix for all subsequent custom schema codes (lookups, columns, supporting entities). Empty string means no prefix is configured."),
```
[the remark this answers is not in the document]
Comment (m-dymytrova) [reply clio/Command/McpServer/Tools/ToolContractGetTool.cs:599]: not confirmed - zero occurrences of `"get-schema-name-prefix"` anywhere in `ToolContractGetTool.cs`. The string the reviewer interprets as a directive is line 597 — this is a response field description inside the `get-app-info` contract. It describes the value of a field in the response body; it does not reference or recommend calling `get-schema-name-prefix`.

[the remark this answers is not in the document]
Comment (m-dymytrova) [reply clio/Command/McpServer/Tools/ToolContractGetTool.cs:599]: registered in `ToolContractCatalog.Contracts` as improvement

File: clio/Command/ApplicationCreateService.cs [hunk 2]
```diff
@@ -315,21 +323,56 @@ private static string SanitizeCode(string code)
...
 		return sanitizedCode;
 	}
 
+	private static void AppendFirstWordWithPrefix(StringBuilder builder, string firstWord, string schemaPrefix)
+	{
+		if (string.IsNullOrEmpty(schemaPrefix))
+		{
+			builder.Append(firstWord);
+			return;
+		}
```
Comment (Alexandr-Kravchuk) [inline clio/Command/ApplicationCreateService.cs:357]: **Blocker: Missing `NormalizeWord` call on the already-prefixed branch**

The original `SanitizeCode` logic called `NormalizeWord(firstWord[3..])` when the first word already carried the `Usr` prefix — stripping non-alphanumeric characters and re-capitalising the first letter. This new method appends `rest` verbatim without normalization.

Inputs that relied on normalization will now silently produce invalid schema codes:
- `"UsrHello-World"` → previously `"UsrHello"` (hyphen stripped), now `"UsrHello-World"` (hyphen retained — invalid Creatio schema name)
- `"Abchello_task"` → previously `"AbcHello"` (underscore stripped, re-capitalised), now `"Abchello_task"` (raw)

Fix: replace the bare `builder.Append(rest)` with `builder.Append(NormalizeWord(rest))`, matching the original behaviour.

File: clio/Command/McpServer/Tools/SchemaNamePrefixTool.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+		[Required]
+		GetSchemaNamePrefixArgs args) {
+		try {
+			ISysSettingsManager sysSettings = commandResolver.Resolve<ISysSettingsManager>(
+				new EnvironmentOptions { Environment = args.EnvironmentName });
+			string rawValue = sysSettings.GetSysSettingValueByCode(SchemaNamePrefixSettingCode);
+			string prefix = rawValue?.Trim().Trim('"') ?? string.Empty;
+			return new SchemaNamePrefixResult(true, prefix);
+		} catch (Exception ex) {
+			return new SchemaNamePrefixResult(false, string.Empty, ex.Message);
```
Comment (Alexandr-Kravchuk) [inline clio/Command/McpServer/Tools/SchemaNamePrefixTool.cs]: **Major (Security): `ex.Message` returned in MCP response may expose sensitive environment details**

Exception messages from network or authentication failures in .NET commonly include target URIs, OAuth endpoint URLs, `client_id` values, Windows credential/NTLM challenge details, and internal topology. Returning `ex.Message` directly in the `SchemaNamePrefixResult.Error` field transmits this over the MCP channel to the AI agent caller — and potentially into agent logs or transcripts.

Fix: map exception types to sanitised, category-based messages before constructing the result:
```csharp
catch (Exception ex) {
    string safeMessage = ex is HttpRequestException or SocketException or TimeoutException
        ? "Network error reading SchemaNamePrefix."
        : ex is UnauthorizedAccessException or AuthenticationException
            ? "Authentication failure reading SchemaNamePrefix."
            : "Unexpected error reading SchemaNamePrefix.";
    // Log ex.ToString() server-side only
    return new SchemaNamePrefixResult(false, string.Empty, safeMessage);
}
```

**Review by Alexandr-Kravchuk — CHANGES_REQUESTED** [review]
## Review: ENG-89314 — clio (PR #579) + ai-driven-app-creation (PR #51)

**Decision: request changes**

Reviewed in conjunction with [ai-driven-app-creation PR #51](https://creatio.ghe.com/engineering/ai-driven-app-creation/pull/51). The feature direction is correct and the design principle (explicit prefix, no silent fallback) is sound. However there is one correctness regression and several gaps that need to be addressed before merge.

---

## Findings

### Blocker

**B1 — `AppendFirstWordWithPrefix` drops `NormalizeWord` on the already-prefixed branch**
`clio/Command/ApplicationCreateService.cs`

The original `SanitizeCode` called `NormalizeWord(firstWord[3..])` when the first word already carried the `Usr` prefix, stripping non-alphanumeric characters and re-capitalising. The new `AppendFirstWordWithPrefix` appends `rest` verbatim, skipping normalization entirely. Any input where the first word already has the active prefix and its remainder contains non-alphanumeric characters or wrong casing will produce an invalid Creatio schema name (e.g., hyphen retained, underscore retained, lowercase preserved). This is a silent regression against previous behaviour and will cause schema-registration failures at the platform level.

Fix: change `builder.Append(rest)` to `builder.Append(NormalizeWord(rest))` in the `firstWord.StartsWith(schemaPrefix)` branch.

---

### Major

**M1 — `ex.Message` returned in MCP response may expose credentials and internal topology** _(inline comment posted)_
`clio/Command/McpServer/Tools/SchemaNamePrefixTool.cs:40`

.NET exception messages from network or auth failures routinely include URIs, `client_id` values, OAuth endpoint URLs, and NTLM/credential details. These travel in the structured `error` field of `SchemaNamePrefixResult` to the AI agent caller and any MCP log sinks. Replace `ex.Message` with sanitised, category-based messages; log the full exception server-side only.

**M2 — `<Prefix>` placeholder cannot be resolved when `create-app` throws on prefix-read failure**
Design gap between PR #51 and PR #579.

PR #51 instructs agents to use `<Prefix>` when the prefix cannot be read at BA time, and to resolve it from the `schema-name-prefix` field in the `create-app` response. PR #579 makes `create-app` throw explicitly when prefix reading fails — meaning no response is returned, so `schema-name-prefix` is never emitted. The resolution path for `<Prefix>` only works when `create-app` succeeds. When it fails, the placeholder is permanently unresolvable and agents have no fallback.

Fix options: (a) return the prefix (or an explicit zero/unknown marker) in the error response from `create-app`, or (b) document explicitly that `<Prefix>` is only deferred (env name unknown at BA time), never unknown (prefix unreadable), and that an unreadable prefix must surface before the Business Plan is produced.

**M3 — `ReadSchemaNamePrefix` and `SchemaNamePrefixSettingCode` duplicated across three files**
`ApplicationCreateService.cs`, `ApplicationInfoService.cs`, `SchemaNamePrefixTool.cs`

The constant `"SchemaNamePrefix"` and the `value?.Trim().Trim('"')` trimming logic are copy-pasted verbatim. If the setting key changes or the quoting convention evolves, all three sites must be updated atomically. The `Trim('"')` rule is a non-obvious business convention that will silently break all three callers if fixed in only one. Extract to a shared static helper (e.g., `SysSettingsHelper.ReadSchemaNamePrefix`).

**M4 — `SchemaNamePrefixTool` uses `commandResolver.Resolve<ISysSettingsManager>` while services use `Func<EnvironmentSettings, ISysSettingsManager>` factory**
`clio/Command/McpServer/Tools/SchemaNamePrefixTool.cs`

The factory registered in `BindingsModule.cs` explicitly branches on OAuth vs basic auth credentials. `commandResolver.Resolve<ISysSettingsManager>` goes through a different resolution path — it is unclear whether it honours the same OAuth/basic branch. If it does not, OAuth-configured environments will silently get an unauthenticated client, fail, and report `success: false`, giving agents a misleading signal. The clio AGENTS.md also states environment-sensitive tools should use the `BaseTool` environment-aware pattern.

Verify that `commandResolver.Resolve<ISysSettingsManager>` routes through the same factory, or align with `BaseTool`.

**M5 — No unit test for `SchemaNamePrefixTool.GetSchemaNamePrefix`**
`clio.tests/Command/McpServer/` — file absent from the diff.

Every other MCP tool in this suite has a unit test. The new tool has only E2E coverage. The clio AGENTS.md policy explicitly states unit tests are necessary but insufficient for MCP changes — E2E-only is also insufficient. Add unit tests covering: successful prefix read, empty prefix, exception propagates as `success: false`, and environment-name validation.

**M6 — E2E assertion only checks `NotBeNull` — actual prefix value never verified**
`clio.mcp.e2e/SchemaNamePrefixToolE2ETests.cs`

`response.SchemaNamePrefix.Should().NotBeNull()` passes for any non-null string, including empty string or a wrong value. The test environment has a known configured prefix. Add `response.SchemaNamePrefix.Should().Be(expectedPrefix)` (or at minimum `NotBeNullOrWhiteSpace`) to verify the reading path end-to-end.

---

### Minor

**m1 — `UsrTaxId` example in AGENTS.md (PR #51) contradicts the new dynamic-prefix rule in the same block**
`AGENTS.md` — Global Invariants

`Semantic Id in business terms is allowed (for example Tax ID -> UsrTaxId)` is in the same paragraph that now instructs agents to use `{activePrefix}`. Agents will follow the example literally and hardcode `Usr` for semantic-Id fields on non-default environments. Change to `{activePrefix}TaxId` with a parenthetical for the default `Usr` case.

**m2 — `schemaPrefix ??= string.Empty` in `SanitizeCode` is dead code**
`clio/Command/ApplicationCreateService.cs`

`ReadSchemaNamePrefix` always returns a non-null string. The null-coalescing guard signals to future readers that null is possible here when it is not. Remove it or document the invariant at the call site.

**m3 — Extra HTTP round-trip per `create-app` and `get-app-info` call, no caching**
`ApplicationCreateService.cs`, `ApplicationInfoService.cs`

`SchemaNamePrefix` is a static system setting that virtually never changes. Each CLI invocation pays one extra blocking remote call. In MCP server mode, sequential calls pay it independently. A process-lifetime or TTL cache keyed by environment URI would eliminate the overhead at negligible complexity cost.

---

## Requirement Coverage

| Requirement | Status |
|---|---|
| Remove silent `"Usr"` fallback — explicit error on read failure | covered |
| New `get-schema-name-prefix` MCP tool | covered |
| `create-app` reads prefix dynamically, returns in response | covered |
| `get-app-info` includes `schema-name-prefix` in response | covered |
| No double-prefix for already-prefixed codes | covered |
| Agent instruction files updated (PR #51) | covered |
| Tests for non-default prefix, empty prefix, digit-start, exception propagation | covered |
| `<Prefix>` placeholder resolution when create-app fails | **missing** (M2 above) |

---

## Confidence Limits

- `SchemaNamePrefixTool` DI resolution path verified by code inspection only; runtime behaviour under OAuth depends on `commandResolver` internals not visible in this diff.
- Performance characteristics of `RemoteDataProvider` (connection reuse, `HttpClient` lifecycle) not fully visible in this diff.
- PR #51 agent instruction files reviewed against the clio PR #579 diff; runtime agent behaviour cannot be mechanically verified.

**Comment by m-dymytrova** [comment]
All findings addressed — changes pushed in commit f0396794.

**B1 — `AppendFirstWordWithPrefix` missing `NormalizeWord`**
Fixed: `builder.Append(NormalizeWord(rest))` in the already-prefixed branch of `ApplicationCreateService.cs`.

**M1 — `ex.Message` exposure in MCP response**
Replaced with three category-mapped catch clauses: `"Network error reading SchemaNamePrefix."`, `"Authentication error reading SchemaNamePrefix."`, `"Failed to read SchemaNamePrefix."` — no raw exception details reach the MCP caller.

**M2 — `<Prefix>` placeholder semantics**
Addressed in the companion ai-driven-app-creation PR #51 (commit 6ecd892): added a clarifying sentence to `context/naming-conventions.md` that `<Prefix>` means the prefix is not yet determined — it is not a substitute for an empty prefix.

**M3 — `ReadSchemaNamePrefix` / `SchemaNamePrefixSettingCode` duplication**
Extracted `SysSettingCodes` static class (`clio/Common/SysSettingCodes.cs`) with the shared constant and a single `ReadSchemaNamePrefix` method containing the `Trim().Trim('"')` logic. All callers (`ApplicationCreateService`, `ApplicationInfoService`, `RemoteEntitySchemaCreator`, `SchemaNamePrefixTool`) delegate to it.

**M4 — `commandResolver.Resolve<ISysSettingsManager>` (unregistered interface)**
Fixed: changed to `commandResolver.Resolve<SysSettingsManager>` (concrete type, which is registered).

**M5 — No unit tests for `SchemaNamePrefixTool`**
Added `clio.tests/Command/McpServer/SchemaNamePrefixToolTests.cs` with 6 tests: stable tool name, success with prefix, empty prefix, network error → `success:false`, unknown env → `success:false`, quote-stripping.

**M6 — E2E assertion too weak**
Strengthened to `NotBeNullOrWhiteSpace` in `SchemaNamePrefixToolE2ETests.cs`.

**m1 — `UsrTaxId` example in AGENTS.md**
Fixed in ai-driven-app-creation PR #51 (commit 6ecd892): updated to `{activePrefix}TaxId, i.e. UsrTaxId when prefix is Usr`.

**m2 — Dead null guard**
Removed `schemaPrefix ??= string.Empty` from `SanitizeCode`.

File: clio/Command/ApplicationCreateService.cs (hunk 2 above)
Comment (m-dymytrova) [reply clio/Command/ApplicationCreateService.cs:357]: Fixed in commit f0396794: changed `builder.Append(rest)` to `builder.Append(NormalizeWord(rest))` in the already-prefixed branch.

File: clio/Command/McpServer/Tools/SchemaNamePrefixTool.cs (hunk 3 above)
Comment (m-dymytrova) [reply clio/Command/McpServer/Tools/SchemaNamePrefixTool.cs]: Fixed in commit f0396794: replaced `ex.Message` with three category-mapped catch clauses — `"Network error reading SchemaNamePrefix."`, `"Authentication error reading SchemaNamePrefix."`, and `"Failed to read SchemaNamePrefix."` — so no raw exception details reach the MCP caller.

**Review by Alexandr-Kravchuk — APPROVED** [review]
## Follow-up review — ENG-89314

All findings from the previous review have been addressed:

- **B1 resolved** — `NormalizeWord(rest)` restored in `AppendFirstWordWithPrefix`; the normalization regression is fixed.
- **M1 resolved** — `ex.Message` replaced with typed, category-based safe messages in `SchemaNamePrefixTool`.
- **M3 resolved** — `SysSettingCodes` helper extracted; constant and trimming logic now in one place, used consistently across `ApplicationCreateService`, `ApplicationInfoService`, `SchemaNamePrefixTool`, and `RemoteEntitySchemaCreator`.
- **M5 resolved** — 6 unit tests added in `SchemaNamePrefixToolTests.cs` covering: prefix read, empty prefix, network failure, unknown environment, quote stripping, and stable tool name.
- **M6 resolved** — E2E assertion updated to `NotBeNullOrWhiteSpace`.

**Remaining minor items (non-blocking):**

- `schemaPrefix ??= string.Empty` in `SanitizeCode` is dead code — `ReadSchemaNamePrefix` never returns null. Safe to remove in a follow-up.
- `SchemaNamePrefixTool` resolves `SysSettingsManager` via `commandResolver.Resolve<SysSettingsManager>` (concrete type) rather than through the `Func<EnvironmentSettings, ISysSettingsManager>` factory used by the service layer. Recommend confirming this path correctly handles OAuth-authenticated environments in a follow-up or integration test.
- No caching for the prefix across sequential MCP tool calls; acceptable for now but worth tracking as more settings go dynamic.

**Review by Alexandr-Kravchuk — APPROVED** [review]
